### ISSUE TITLE
Redirect after login

### ISSUE THREAD

**Issue by MichaelBlume** [Marri/glowfic#132]
If I try to go to, say, my unread page, but it turns out I'm not logged in, I'll be directed to the login page. After I log in, it'd be nice to be taken automatically back to the unread page.

**Comment by Throne3d** [comment]
I don't think any redirects presently direct you to the login page? The code looks like it should direct you to `root_path`, so it could be hard to then redirect you back to the page you wanted assuming you do then log in, and it shouldn't persist in the session as I'd find it extremely weird to load a page, be redirected, then load the login and be directed to the original page I loaded.

If we altered the code to direct you to the login page, instead of `root_path`, then I think it'd make sense for this to be added (probably as a 'forward_to' parameter or something), but that's probably – in my opinion – a prerequisite to this.

**Comment by dhouck** [comment]
I would agree with this being weird if there were anything on the root besides a two-sentence welcome message (and, in this use case, an error message). As it is, there is no reason to want to be on that screen in particular, so instead redirecting to `/&forward_on_login=/posts/unread` or something seems like it would be sensible.

**Comment by Throne3d** [comment]
That's not particularly my point, or perhaps you got my point but responded obliquely with the first part? We don't currently redirect you (with an actual redirect as opposed to a verbal redirect) to the login page at all, so it'd be bizarre to start directing you from the login page to other places based on your history imo, as it'd be at least two steps removed.

We can probably also handle the login redirect without a parameter by using the HTTP `Referrer` header. I'm not sure why we would redirect to root with that parameter anyway (`/?…`) and so I assume that was a typo and either the slash was not meant to be present or you meant `/login?…`.

**Comment by dhouck** [comment]
I did not know there was an actual login page; there doesn’t seem to be any link to it from anywhere.

When I type https://glowfic.com/posts/unread in the address bar (or more likely, refresh that page with the current login cookie issues) and I am not logged in, I find myself on https://glowfic.com/, which has login boxes. I log in from there, and I find myself on https://glowfic.com/ while logged in. My expected workflow would be going to https://glowfic.com/posts/unread, seeing a page with login boxes, putting information in the login boxes and hitting “Log in”, and finding myself back on https://glowfic.com/posts/unread; as an end-user I don’t care what the URL of the logged-out page with login boxes is.

(for comparison, if I type go to https://glowfic.com/users/31 while logged out, and I want to be logged in, I fill in my login information and hit the “Log in” button, and I am still on https://glowfic.com/users/31).

I suggested redirect to root with parameter because the current behavior is redirect to root without parameter, and I didn’t think of non-parameter ways of redirect after reply. Since there is a login page, redirecting there with parameter makes more sense. I don’t think `Referrer` would work, because the `Referrer` when hitting the login button is whatever page had a login button, not the page I was on before that.

**Comment by Throne3d** [comment]
I understand the expected workflow but hadn't taken into account the "there are login buttons on the root page in the header as everywhere else" part of this. There is in fact a login page, mostly used I think just on mobile devices where the form in the header was inconveniently large.

I personally think using HTTP `Referrer` would be *neater*, but it has potential drawbacks as "the user got to the login page and then decided they wanted to log in but not return to the page they were on previously".

We also need to consider how this will interact with users sending `POST` requests who are then redirected to the login page; presumably logging in should not then resend the `POST` (because that sounds awfully destructive and like it could have large negative side effects and also like it'd be hard to enable) but then where should they be sent instead?

**Comment by dhouck** [comment]
The problem with using `Referrer` is that if I have a failed login attempt, the `Referrer` information is lost. I suppose it could convert it to a URL parameter then, but if there’s going to be a fallback of a URL parameter it might as well just always be one.

If you do use `Referrer`, I would expect any non-Glowfic `Referrer`, or a `Referrer` of the login page, to mean it just redirects to the root page on login (or Continuities, as is the current behavior, or an account option).

**Comment by Throne3d** [comment]
*Shrug*. We could get around the former issue by saving the referrer data. We'd have to do that with a URL parameter on a failed login attempt anyway. I'm not sure which would be more effort.

Agreed on the other point.
